ScoutSuite/providers/base/configs/resources.py

```diff
@@ -0,0 +1,10 @@
+# -*- coding: utf-8 -*-
+
+import abc
+
+
+class Resources(dict, metaclass=abc.ABCMeta):
```

Comment:
Just saw the comments on the AWS Lambda PR about removing the `Config` suffix; I'm ok with it as long as we're consistent throughout the services/providers, but I still prefer having it as I think it's clearer as to what it represents. I'm not a fan of classes having a plural name though (ie. `Resources`). In my opinion, each instance should represent one config of that given resource and not multiple ones so the name should reflect that.

Reply:
I think most of us said we were confused and annoyed by the `Config` suffix. I'm not sure it's a clearer convention. I wouldn't be shocked if I entered the codebase for the first time and saw a `Lambda` object containing a bunch of info about a `Lambda`. I would expect `LambdaConfig` to be a mutable object, perhaps provided by the user, which affects the behavior of the software, as it is often the case with `_______Config` objects.

I agree `Resources` isn't an ideal name (I think I left a comment about that in another PR actually), but I don't agree with saying
> each instance should represent one config of that given resource

The whole idea of our refactoring is to build a hierarchical structure. If each instance represents one resource, then our hierarchy can only have one level of depth. Also, most of the API calls are designed to retreive a batch of resources (in a given region or project, for example). If each instance represents one resource, who does the API call?

That being said, I'm not sure `Resource` is the right name either, because it implies the class contains only one resource, which might not be true. Did I just loop back to `Config` then? No, still think it's a bad name, but maybe something more like that.

Reply:
> I think most of us said we were confused and annoyed by the `Config` suffix. I'm not sure it's a clearer convention. I wouldn't be shocked if I entered the codebase for the first time and saw a `Lambda` object containing a bunch of info about a `Lambda`. I would expect `LambdaConfig` to be a mutable object, perhaps provided by the user, which affects the behavior of the software, as it is often the case with `_______Config` objects.

I agree with that.

> Also, most of the API calls are designed to retreive a batch of resources (in a given region or project, for example).
> I'm not sure `Resource` is the right name either, because it implies the class contains only one resource, which might not be true.

That's why I suggested `Resources` over `Resource` (and it's just details but we we're used to the method `fetch_all`... whose name implies several resources and not only one).

Reply:
You're right, actually, I hadn't thought about `fetch_all`. Also, we don't really have explicit classes for the leaves. At least, in my refactoring of EC2, I didn't need any.

Reply:
> I think most of us said we were confused and annoyed by the `Config` suffix. I'm not sure it's a clearer convention. I wouldn't be shocked if I entered the codebase for the first time and saw a `Lambda` object containing a bunch of info about a `Lambda`. I would expect `LambdaConfig` to be a mutable object, perhaps provided by the user, which affects the behavior of the software, as it is often the case with `_______Config` objects.

Fair enough!

> If each instance represents one resource, who does the API call?

I believe we discussed this when designing the refactoring, the parent config would be the one fetching the child resource. ie. `Kubernetes` would fetch the clusters and create instances of `Cluster` which would take care of parsing the necessary info for the rules. 

The problem I see with having the class fetch all the instances of its own resource type is that the necessary info to fetch them can vary  depending on the context in which those resources exist. For example, compute instances, networks and disks are used both by Kubernetes Engine and Compute Engine, but they are fetched with different parameters depending on which service you are using. 

In the design I'm describing, you would have `Kubernetes` and `GCE` classes which would have multiple `Instance`, `Disk` and `Networks` objects as children. This would prevent code duplication for resources that are used by different services.

> (and it's just details but we we're used to the method `fetch_all`... whose name implies several resources and not only one).

In my opinion, the current name isn't ideal. I would prefer something like `parse()` which would get the info used by rules for that given resource and fetch its children resources.

Reply:
> In the design I'm describing, you would have `Kubernetes` and `GCE` classes which would have multiple `Instance`, `Disk` and `Networks` objects as children. This would prevent code duplication for resources that are used by different services.

How do you mean? I'm not sure I understand how it would prevent code duplication. If you're talking about the data fetching code, it is now wrapped in a facade (or injected as a service, but wrapped somehow), so there is no code duplication there. The rest of the code is just parsing, which is specific to each parent class, and some kind of loop for parsing every resource, which can be generalized. 

The difference I see between what you propose and what we (or at least I, but I believe that's mostly what @misg did, I'd have to re-review his PR) started implementing is simply moving the calls to the facade and the parsing method up a level. 

This is what I implemented in my PR.
![image](https://user-images.githubusercontent.com/6698529/53306378-5ffae100-385a-11e9-80c5-f83535352eee.png)

Reply:
Yes, I am indeed suggesting moving the call to the facade one level, since getting resources of the same type for different services might require different calls. This way, the parent wouldn't pass regions but rather/additionally a list of raw resources, allowing for let's say GKE and GCE to get Disk resources in a different manner but using the same `Disks` class to parse the raw disks.

Reply:
> For example, compute instances, networks and disks are used both by Kubernetes Engine and Compute Engine, but they are fetched with different parameters depending on which service you are using.

This is currently handled through the provider's `preprocessing()` method, which allows for cross-service data processing/normalization.

Reply:
> In my opinion, the current name isn't ideal.

The legacy naming convention (`go_to_and_do`, `fetch_all`, etc.) is quite confusing indeed.

Reply:
>> In my opinion, the current name isn't ideal.
>
>The legacy naming convention (`go_to_and_do`, `fetch_all`, etc.) is quite confusing indeed.

You forgot `new_go_to_and_do`.

> Yes, I am indeed suggesting moving the call to the facade one level, since getting resources of the same type for different services might require different calls.

So effectively, this means the `EC2` class would get the `Vcps`, and the `Vpcs` class would get the. instances. That's a bit counter-intuitive, I feel like each resource should be responsible of fetching its own data. :/ After, if we want to move the parsing inside another class, we could, but it's just more code to write.